Added changes to enable full file cache stats

mayanksharma27 opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
- Added Changes to enable FullFileCacheStats
- Added integration tests for FileCacheStats and FullFileCacheStats for writable warm.

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/17479.
Meta Issue https://github.com/opensearch-project/OpenSearch/issues/13149

### Check List
- [ ✓ ] Functionality includes testing.
- [ NA] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [NA] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

gbbafna commented on server/src/main/java/org/opensearch/index/store/remote/filecache/FileCache.java:
```diff
@@ -51,12 +53,66 @@
+        this.circuitBreaker = circuitBreaker;
+        this.fullFileCacheStatsCollector = new FullFileCacheStatsCollector();
+
+        this.theCache = SegmentedCache.<Path, CachedIndexInput>builder()
+            // use length in bytes as the weight of the file item
+            .weigher(CachedIndexInput::length)
+            .listener((removalNotification) -> {
+                RemovalReason removalReason = removalNotification.getRemovalReason();
+                CachedIndexInput value = removalNotification.getValue();
+                Path key = removalNotification.getKey();
+
+                if (removalReason != RemovalReason.REPLACED) {
+
+                    // check if full file and publish metric
+                    if (value instanceof CachedFullFileIndexInput) {
+                        fullFileCacheStatsCollector.recordUsedFileBytes(value.length(), false);
+                    }
+
+                    catchAsRuntimeException(value::close);
+                    catchAsRuntimeException(() -> Files.deleteIfExists(key));
+                }
+            })
+            .capacity(capacity)
+            .build();
+        ;
+
```
Lets de-dupe the code which is there in other cxtor as well .

rayshrey replied:
+1 on deduping the code in other constructor.
Also in my opinion it is better to make modifications to the stats at the base LRUCache level instead of FileCache level since all other stats are being modified there only.

mayanksharma27 replied:
Done.

gbbafna commented on server/src/test/java/org/opensearch/index/store/remote/filecache/FileCacheTests.java:
```diff
@@ -77,6 +79,35 @@ public void testGet() {
+            );
+            fileCache.put(filePath, new CachedFullFileIndexInput(fileCache, filePath, fileCachedIndexInput));
+        }
+        // verify all blocks are put into file cache
```
nit : `all files` ?

gbbafna replied:
Can we add eviction test as well here ?

mayanksharma27 replied:
Done

rayshrey reviewed (commented):
Looks good to me from the functionality point of view.
But have a few concerns related to extensibility (specially when we are going to introduce the pinning functionality)

- We have added FullFileCacheStats to account for the stats of the full file. Will we require a similar file for accounting for stats of blocked files (and pinned files in future) ? In that case should we not have it in a single place - [CacheUsage](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/store/remote/utils/cache/CacheUsage.java) would be a good place for that since we are already maintaning the usage and activeUsage overall there.

- Also adding to my comment on making changes in the LRUCache instead of FileCache, how about adding changes to the [Node](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/store/remote/utils/cache/LRUCache.java#L75) in LRUCache to check whether the IndexInput is for block or full file. We already have a weight param for Node - we can add file type, pinned status etc as per our use case to simplify the stats/usage aggregation. We would just need to make changes in the replaceNode and addNode functions to account for our needed stats/usages. Something similar to below.

```
static class Node<K, V> {
        final K key;
        V value;
        long weight;
        int refCount;
        // Add below properties for Node
        FileType fileType; 
        PinStatus pinned;
}
```

```
public class CacheUsage {
        private final long usage;
        private final long activeUsage;
        // Add below properties in CacheUsage
        private final long blockUsage;
        private final long nonBlockUsage;
        private final long pinnedUsage;
}
```

@gbbafna @ankitkala @mayanksharma27 thoughts on the above approach ?

mayanksharma27 commented on server/src/internalClusterTest/java/org/opensearch/remotestore/WritableWarmIT.java:
```diff
@@ -168,4 +177,82 @@ public void testWritableWarmBasic() throws Exception {
+
+        FileCache fileCache = internalTestCluster.getDataNodeInstance(Node.class).fileCache();
+
+        // TODO: Make these validation more robust.
```
Done

mayanksharma27 commented on server/src/main/java/org/opensearch/index/store/remote/filecache/FullFileCacheStats.java:
```diff
@@ -0,0 +1,135 @@
+/**
+ * Statistics on full file cache.
+ */
+@PublicApi(since = "3.0.0")
```
Done

mayanksharma27 commented on server/src/main/java/org/opensearch/index/store/remote/filecache/FullFileCacheStatsCollector.java:
```diff
@@ -0,0 +1,70 @@
+ */
+@PublicApi(since = "3.0.0")
+public class FullFileCacheStatsCollector {
+    private final CounterMetric activeFullFileBytes = new CounterMetric();
+    private final CounterMetric activeFullFileCount = new CounterMetric();
```
ack.

mayanksharma27 commented:
> * We have added FullFileCacheStats to account for the stats of the full file. Will we require a similar file for accounting for stats of blocked files (and pinned files in future) ? In that case should we not have it in a single place - [CacheUsage](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/store/remote/utils/cache/CacheUsage.java) would be a good place for that since we are already maintaning the usage and activeUsage overall there.
  -  We wont be requiring a similar file for both of these as current `filecachestats` would handle it. As part of the recent commit I have -
    -  Moved `Usage` and `ActiveUsage` as part of stats itself; and removed CacheUsage altogether; this is done to keep FileCacheStats as single source of truth for all metrics related to LRU Cache.


> * Also adding to my comment on making changes in the LRUCache instead of FileCache, how about adding changes to the [Node](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/store/remote/utils/cache/LRUCache.java#L75) in LRUCache to check whether the IndexInput is for block or full file. We already have a weight param for Node - we can add file type, pinned status etc as per our use case to simplify the stats/usage aggregation. We would just need to make changes in the replaceNode and addNode functions to account for our needed stats/usages. Something similar to below.

  - While adding pinning info in Node does makes sense, adding business logic - like filetype needed just to publish metrics and not needed for cache functionality is not clean. I instead have added this logic as part of CacheStats class.

skumawat2025 commented on server/src/main/java/org/opensearch/index/store/remote/utils/cache/stats/DefaultStatsCounter.java:
```diff
@@ -43,29 +76,121 @@ public void recordMisses(K key, int count) {
         return snapshot().toString();
     }
+
+    private boolean isFullFile(V value) {
```
We are using this type checking every-time we remove, add, get, and put value to FileCache. 
Would it make sense to have this value pre-computed for FullFIleCachedIndexInput? 
Maybe introducing a method in `CachedIndexInput` Thoughts?

gbbafna replied:
That is a good point @skumawat2025 . So for every get we would be calling `instanceof` . We should do a benchmark for this change by reusing `FileCacheBenchmark` . If we see any perf issues, we will need to solve for it .

mayanksharma27 replied:
Ran the benchmarking with and without instanceof changes, and the results are almost similar. 


**Before changes BenchMarking Results**


```
REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
experiments, perform baseline and negative tests that provide experimental control, make sure
the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
Do not assume the numbers tell you what you want them to tell.


NOTE: Current JVM experimentally supports Compiler Blackholes, and they are in use. Please exercise
extra caution when trusting the results, look into the generated code to check the benchmark still
works, and factor in a small probability of new VM bugs. Additionally, while comparisons between
different JVMs are already problematic, the performance difference caused by different Blackhole
modes can be very significant. Please make sure you use the consistent Blackhole mode for comparisons.


Benchmark (concurrencyLevel) (maximumNumberOfEntries) Mode Cnt Score Error Units
FullCacheBenchmark.get 1 65536 thrpt 5696.576 ops/ms
FullCacheBenchmark.get 1 1048576 thrpt 1893.665 ops/ms
FullCacheBenchmark.get 8 65536 thrpt 10201.958 ops/ms
FullCacheBenchmark.get 8 1048576 thrpt 4060.831 ops/ms
FullCacheBenchmark.put 1 65536 thrpt 5280.068 ops/ms
FullCacheBenchmark.put 1 1048576 thrpt 2247.362 ops/ms
FullCacheBenchmark.put 8 65536 thrpt 10601.330 ops/ms
FullCacheBenchmark.put 8 1048576 thrpt 4249.552 ops/ms
FullCacheBenchmark.remove 1 65536 thrpt 18369.760 ops/ms
FullCacheBenchmark.remove 1 1048576 thrpt 16386.680 ops/ms
FullCacheBenchmark.remove 8 65536 thrpt 26378.060 ops/ms
FullCacheBenchmark.remove 8 1048576 thrpt 20196.221 ops/ms
FullCacheBenchmark.replace 1 65536 thrpt 5202.601 ops/ms
FullCacheBenchmark.replace 1 1048576 thrpt 1965.434 ops/ms
FullCacheBenchmark.replace 8 65536 thrpt 11147.688 ops/ms
FullCacheBenchmark.replace 8 1048576 thrpt 4248.442 ops/ms


[Incubating] Problems report is available at: file:///Users/smynk/opensource/OpenSearch-mayank-clone/OpenSearch/build/reports/problems/problems-report.html


BUILD SUCCESSFUL in 5m 32s
44 actionable tasks: 2 executed, 42 up-to-date
Watched directory hierarchies: [/Users/smynk/opensource/OpenSearch-mayank-clone/OpenSearch, /Users/smynk/opensource/OpenSearch-mayank-clone2/OpenSearch]

```



** After changes (with instance of) BenchMarking Results**



```
REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
experiments, perform baseline and negative tests that provide experimental control, make sure
the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
Do not assume the numbers tell you what you want them to tell.


NOTE: Current JVM experimentally supports Compiler Blackholes, and they are in use. Please exercise
extra caution when trusting the results, look into the generated code to check the benchmark still
works, and factor in a small probability of new VM bugs. Additionally, while comparisons between
different JVMs are already problematic, the performance difference caused by different Blackhole
modes can be very significant. Please make sure you use the consistent Blackhole mode for comparisons.


Benchmark (concurrencyLevel) (maximumNumberOfEntries) Mode Cnt Score Error Units
FullCacheBenchmark.get 1 65536 thrpt 5642.441 ops/ms
FullCacheBenchmark.get 1 1048576 thrpt 1858.423 ops/ms
FullCacheBenchmark.get 8 65536 thrpt 9190.302 ops/ms
FullCacheBenchmark.get 8 1048576 thrpt 4149.956 ops/ms
FullCacheBenchmark.put 1 65536 thrpt 4847.883 ops/ms
FullCacheBenchmark.put 1 1048576 thrpt 2140.148 ops/ms
FullCacheBenchmark.put 8 65536 thrpt 9406.632 ops/ms
FullCacheBenchmark.put 8 1048576 thrpt 4446.392 ops/ms
FullCacheBenchmark.remove 1 65536 thrpt 21195.795 ops/ms
FullCacheBenchmark.remove 1 1048576 thrpt 15670.369 ops/ms
FullCacheBenchmark.remove 8 65536 thrpt 27016.226 ops/ms
FullCacheBenchmark.remove 8 1048576 thrpt 20154.585 ops/ms
FullCacheBenchmark.replace 1 65536 thrpt 5564.645 ops/ms
FullCacheBenchmark.replace 1 1048576 thrpt 1819.900 ops/ms
FullCacheBenchmark.replace 8 65536 thrpt 10835.397 ops/ms
FullCacheBenchmark.replace 8 1048576 thrpt 4170.602 ops/ms



```

skumawat2025 commented on server/src/main/java/org/opensearch/index/store/remote/utils/cache/stats/DefaultStatsCounter.java:
```diff
@@ -8,19 +8,40 @@
  * @opensearch.internal
  */
-public class DefaultStatsCounter<K> implements StatsCounter<K> {
+public class DefaultStatsCounter<K, V> implements StatsCounter<K, V> {
```
How about having a different implementation of StatsCounter in case of FullFileStats? Maybe using a Factory while initialising Cache. Currently FullFileStats won't make sense in case of SearchableSnapshots right?

gbbafna replied:
It can make sense if we support pinning some files there.

rayshrey replied:
Same suggestion here as well - have a base stats counter implementation and then have DefaultStatsCounter as below
```
DefaultStatsCounter {
    StatsCounter overallStatsCounter;
    StatsCounter fullFileStatsCounter;
    StatsCounter blockedFileStatsCounter;
    StatsCounter pinnedFileStatsCounter;
}
```

skumawat2025 commented on server/src/main/java/org/opensearch/index/store/remote/utils/cache/stats/CacheStats.java:
```diff
@@ -27,6 +28,144 @@ public final class CacheStats {
+     * Represents Stats about FullFiles in a {@link RefCountedCache}.
+     */
+    @ExperimentalApi
+    public class FullFileStats {
```
Can be static class?

rayshrey replied:
Similar recommendation as provided for FileCacheStats, let's use inheritance to make things extensible here as well (Have a base class with all common fields and then use those classes together in a single class.

rayshrey commented on server/src/main/java/org/opensearch/index/store/remote/filecache/FileCache.java:
```diff
@@ -145,8 +149,8 @@ public CacheStats stats() {
-        CacheUsage cacheUsage = theCache.usage();
-        logger.trace("Total Usage: " + cacheUsage.usage() + " , Active Usage: " + cacheUsage.activeUsage());
+        long cacheUsage = theCache.usage();
+        logger.trace("Total Usage: " + cacheUsage);
```
Can we log the active usage here as well since it was logged previously as well.

mayanksharma27 replied:
done

rayshrey commented on server/src/main/java/org/opensearch/index/store/remote/utils/cache/stats/DefaultStatsCounter.java:
```diff
@@ -43,29 +76,121 @@ public void recordMisses(K key, int count) {
     }
+
+    private boolean isFullFile(V value) {
+        return value instanceof CachedFullFileIndexInput;
```
Instead of using this instanceof check, I think it will be better to use the Path of the the file to identify it's type(block files will have block as part of their name). This will also ensure that we don't need to change the signature of StatsCounter<K> to StatsCounter<K,V>, thus minimising the changes.

rayshrey commented on server/src/main/java/org/opensearch/index/store/remote/filecache/FileCacheStats.java:
```diff
@@ -33,6 +41,7 @@ public class FileCacheStats implements Writeable, ToXContentFragment {
     private final long evicted;
     private final long hits;
     private final long misses;
+    private final FullFileCacheStats fullFileCacheStats;
```
This approach is not extensible, in case we need to add another BlockedFileCacheStats or PinnedFileCacheStats, we would have to create multiple duplicate classes with the same fields and functionality. Let's create a base class similar to below 
```
public class FileCacheStatsBase {
    private long used;
    private long total;
    private long evicted;
    private long hits;
    private long misses;
}
```

Then we can create overallCacheStats, FullFileCacheStats, BlockedFileCacheStats, PinnedFileCacheStats and have our FileCacheStats as below

```
public class FileCacheStats {
    private FileCacheStatsBase overallCacheStats;
    private FileCacheStatsBase FullFileCacheStats;
    private FileCacheStatsBase BlockedFileCacheStats;
    private FileCacheStatsBase PinnedFileCacheStats;

    // return the overall stats for currently existing methods such as getHits/getMisses etc
    public long getCacheHits/getCacheMisses/...() {
        overallCacheStats.hits/misses;
    }

    // similarly for all other stats
    public long getFullFileCacheStats() {
        return FullFileCacheStats;
    }

}
```

rayshrey commented on server/src/main/java/org/opensearch/index/store/remote/utils/cache/stats/IRefCountedCacheStats.java:
```diff
@@ -0,0 +1,124 @@
+ *  Statistics about the performance of a {@link RefCountedCache}.
+ */
+@ExperimentalApi
+public interface IRefCountedCacheStats {
```
Nit: If no one apart from RefCountedCacheStats is using it, we can skip creating this interface.

In case we want to keep it, let's change the interface name to RefCountedCacheStats and the concrete class to RefCountedCacheStatsImpl or something intuitive.

mayanksharma27 replied:
Both `AggregateRefCountedCacheStats` and `RefCountedCacheStats` use this interface. Keeping this interface named as is.

mayanksharma27 commented on server/src/main/java/org/opensearch/index/store/remote/filecache/AggregateFileCacheStats.java:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
```
WriteTo Method is called by NodeStats WriteTo Method. Similarly toXContent is called by NodeStats toXContent method see links below.

-  https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/admin/cluster/node/stats/NodeStats.java#L521
- https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/admin/cluster/node/stats/NodeStats.java#L629

I had earlier only added overall stats in toXContent,  have updated toXContent to be in sync with WriteTo.

mayanksharma27 commented on server/src/main/java/org/opensearch/index/store/remote/filecache/FileCache.java:
```diff
@@ -208,19 +214,40 @@ public void restoreFromDirectory(List<Path> fileCacheDataPaths) {
-            stats.missCount()
+            new FileCacheStats(
+                overallCacheStats.activeUsage(),
+                0L,
```
Fixed.

mayanksharma27 commented on server/src/main/java/org/opensearch/index/store/remote/filecache/FileCacheStats.java:
```diff
@@ -113,38 +104,27 @@ public long getCacheMisses() {
     static final class Fields {
-        static final String FILE_CACHE = "file_cache";
-        static final String TIMESTAMP = "timestamp";
+        static final String FULL_FILE_STATS = "full_file_stats";
```
updated the string to `file_stats `. We dont need a constructor parameter here, as the key in the aggregate stats class is distinctive.

mayanksharma27 commented on server/src/main/java/org/opensearch/index/store/remote/filecache/AggregateFileCacheStats.java:
```diff
@@ -0,0 +1,178 @@
+ *
+ * @opensearch.api
+ */
+@PublicApi(since = "2.7.0")
```
Sure

mayanksharma27 commented on server/src/internalClusterTest/java/org/opensearch/remotestore/WritableWarmIT.java:
```diff
@@ -172,4 +179,88 @@ public void testWritableWarmBasic() throws Exception {
+            fail("File Cache Stats should not be null");
+        }
+
+        // FileCacheStats fullFileCacheStats = fileCacheStats.;
+        //
+        // if (Objects.isNull(fullFileCacheStats)) {
+        // fail("Full File Cache Stats should not be null");
+        // }
```
ack

mayanksharma27 commented on CHANGELOG.md:
```diff
@@ -54,7 +55,6 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 ### Deprecated
 
 ### Removed
-- [repository-s3] Removed existing ineffective `server_side_encryption` setting ([#18312](https://github.com/opensearch-project/OpenSearch/pull/18312))
```
fixed unintentional changes

rayshrey reviewed (approved):
LGTM
